### PR TITLE
[SPARK-10542] [PYSPARK] fix serialize namedtuple

### PR DESCRIPTION
**Pull request by davies**

### CODE REVIEW

**Comment by JoshRosen** [comment]
Over at https://issues.apache.org/jira/browse/SPARK-10544, someone commented to mention that other types of built-in types do not seem to be pickleable in 1.5.  For instance, here's the example that they gave:

```python
sc.parallelize(["the red", "Fox Runs", "FAST"]).map(str.lower).count()
```

However, this specific example also seems to fail in 1.3.1, so I don't think that this is a regression. Just wanted to mention this discussion here to make sure you were aware of it.

**Comment by JoshRosen** [comment]
Do you have any intuition for why this worked prior to 1.5 without the changes implemented here?

File: python/pyspark/tests.py [hunk 1]
```diff
@@ -218,6 +218,11 @@ def test_namedtuple(self):
         p2 = loads(dumps(p1, 2))
         self.assertEqual(p1, p2)
 
+        from pyspark.cloudpickle import dumps
+        P2 = loads(dumps(P))
+        p3 = P2(1, 3)
+        self.assertEqual(p1, p3)
```
Comment (JoshRosen) [inline python/pyspark/tests.py]: Ah, interesting: presumably `P` and `P2` are different classes but instances created from them are still comparable for equality. Do we also need to check that those instances claim to belong to the same class? It seems way less likely that users could rely on the class comparison behavior, so probably not a huge priority to look at.

**Comment by JoshRosen** [comment]
Actually, one point of confusion: it looks like `python/pyspark/serializers.py` had some methods for serializing namedtuple classes with its `_restore` and `_hack_namedtuple(cls)` methods, but this patch also adds logic to `cloudpickle.py` to handle pickling of these instances. Do we need both of these methods? Are some only needed by certain versions of Python? It would be good to add some comments to explain some of this stuff, since it's not obvious from reading the code.

**Comment by davies** [comment]
The HACK in serializers.py is used for cPickler, not cloudpickle.

**Comment by davies** [comment]
Before 1.5, the old way work in CPython, but not PyPy (we don't have a unit test for it).

File: python/pyspark/tests.py (hunk 1 above)
Comment (davies) [inline python/pyspark/tests.py]: These instances should become to difference classes.

**Comment by davies** [comment]
@JoshRosen BTW, this patch introduce a special case for namedtuple, it should be safe to merge into branch-1.5.

**Comment by JoshRosen** [comment]
Empirically, this seems to work, so unless you think that we should investigate the root cause any further I'm fine with giving this an LGTM and merging to 1.5. Feel free to merge yourself, or I can do it.

**Comment by davies** [comment]
I tried to find the root cause, but it seems hard to work in all Python versions (you can see them in the older commit), finally switch to current approach. merging this into master and 1.5 branch, thanks!

**Comment by coderfi** [comment]
FYI, this works for us @ [NinthDecimal](http://www.ninthdecimal.com). Thanks for the fix, it was a stumper!

Python 2.7.5
Spark-1.5 Branch @ [4c4a9ba28d9052fad45caca9a1eba9ef9db309d5](https://github.com/apache/spark/tree/4c4a9ba28d9052fad45caca9a1eba9ef9db309d5)

**Comment by silentsnooc** [comment]
Hi! I am not sure if this is related but is I look for this issue everything points me here basically. I'm getting 

```
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python2.7/pickle.py", line 649, in save_dict
    self._batch_setitems(obj.iteritems())
  File "/usr/lib/python2.7/pickle.py", line 681, in _batch_setitems
    save(v)
  File "/usr/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle.py", line 315, in save_builtin_function
    return self.save_function(obj)
  File "/opt/spark/python/lib/pyspark.zip/pyspark/cloudpickle.py", line 191, in save_function
    if islambda(obj) or obj.__code__.co_filename == '<stdin>' or themodule is None:
AttributeError: 'builtin_function_or_method' object has no attribute '__code__'
```

When trying to create a data frame from an RDD:

```python
        rdd = self.sc.textFile(self.input_file_path).map(lambda line: self.process_line(line))

        schema = StructType([StructField(u'Variable', StringType(), nullable=False),
                             StructField(u'Time', TimestampType(), nullable=False),
                             StructField(u'Value', FloatType(), nullable=False)])

        return sql_context.createDataFrame(rdd, schema)
```

I am on PySpark 1.6.0 - any ideas what I'm doing wrong here?

**Comment by philastrophist** [comment]
I also get this error when using namedtuples

```
n = namedtuple('test', ['a', 'b'])
cloudpickle.dumps(n(1,2))
```

**Pull request closed without merging**